### PR TITLE
Converted project to Typescript

### PR DESCRIPTION
**Pull request by jhsware**
IMPORTANT: I don't know how to properly test the functionality of this apart from running tests.

TODO: Perform end-to-end tests

### CODE REVIEW

File: src/lib/utils.ts [hunk 1]
```diff
@@ -49,7 +49,7 @@ function createStartDate(round) {
   return `${startDate}T06:00:00Z`;
 }
 
-module.exports = {
+export {
```
Comment (me97esn) [inline src/lib/utils.ts:52]: Question: why go from commonjs modules to es modules in this PR? Is there a reason why this is done in a typescript PR? Don't typescript support commonjs, or is there another reason (as far as I know you can use ES modules in node without using typescript, so why haven't we been using them before and why are we using them now)?

File: src/lib/ug.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@ async function ldapUnbind() {
 
 async function ldapSearch({
   base = "OU=UG,DC=ug,DC=kth,DC=se",
-  filter = "",
+  EqualityFilter: filter = "",
```
Comment (me97esn) [inline src/lib/ug.ts:22]: I realize that I didn't understand this syntax before this PR, and even less after this PR :)
Is this an object with default parameters? And what does line 22 in this PR mean? Can you please explain, I'm confused :confused:

File: src/lib/utils.ts (hunk 1 above)
Comment (jhsware) [reply src/lib/utils.ts:52]: The default import in TS is import ..., commonjs require() style imports are only supported for compatibility.

File: src/lib/ug.ts [hunk 3]
```diff
@@ -87,9 +87,4 @@ async function loadMembers(groupName) {
   return (await getUsersForMembers(members)).map((user) => user.ugKthid);
 }
 
-module.exports = {
-  ldapClient,
-  ldapBind,
-  ldapUnbind,
-  loadMembers,
-};
+export { ldapBind, ldapUnbind, loadMembers };
```
Comment (me97esn) [inline src/lib/ug.ts:90]: Here we export one less property then before. Is this on purpose?

File: src/lib/ug.ts (hunk 2 above)
Comment (jhsware) [reply src/lib/ug.ts:22]: `EqualityFilter:` specifies the type of parameter `filter` in argument object, so it is there to tell ldapSearch() what we are passing. Probably something ldapSearch() required. Think of it as a hint to know that this isn't a random string. It is possible that the EqualityFilter has specific requirements, but it could also just be a typing hint. I didn't investigate.

File: src/lib/courseRoundsUtils.ts [hunk 4]
```diff
@@ -95,11 +96,12 @@ function isFarFuture(round) {
 function shouldHaveAntagna(round) {
   const THREE_DAYS = 3 * 24 * 60 * 60 * 1000;
   const startDate = new Date(createStartDate(round));
+  const now = new Date();
 
-  return new Date() - startDate < THREE_DAYS;
+  return now.valueOf() - startDate.valueOf() < THREE_DAYS;
```
Comment (me97esn) [inline src/lib/courseRoundsUtils.ts:101]: Why is this changed in this PR? Shouldn't we keep the original functionality here and _just_ add Typescript support?

File: src/lib/ug.ts (hunk 3 above)
Comment (jhsware) [reply src/lib/ug.ts:90]: ldapClient wasn't used in the project and is set at runtime after it was exported, so I believe it will always be undefined.

File: src/lib/kopps.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
+/* eslint-disable import/prefer-default-export */
+import got from "got";
+import log from "skog";
+
+class KoppsError extends Error {
```
[the remark this answers is not in the document]
Comment (jhsware) [reply src/lib/kopps.ts:5]: Sorry about that!

File: src/lib/ug.ts (hunk 2 above)
Comment (me97esn) [reply src/lib/ug.ts:22]: My first question was more about the syntax. What does an object with equal signs mean? I haven't seen it before.

The second question is about the _change_ of the filter, is it a good idea to change this _in a typescript_ PR?

File: src/lib/utils.ts (hunk 1 above)
Comment (me97esn) [reply src/lib/utils.ts:52]: So if we were to go from `require` to `import` in another PR, without using Typescript, and then create a PR that _only_ introduces Typescript, what would the changes be? I'm trying to see the value this PR brings to the project.

File: src/lib/courseRoundsUtils.ts (hunk 4 above)
Comment (jhsware) [reply src/lib/courseRoundsUtils.ts:101]: new Date() - startDate is converted by the JS-engine to (new Date()).valueOf() - startDate.valueOf(). I think the reason it ended up like that was due to some debugging that caused an intermediary state where this made a lot more sense. But I can't remember. I am pretty sure I didn't change it for change sake...😅

Comment (me97esn) [reply src/lib/courseRoundsUtils.ts:101]: Also note that this date functionality has been changed in the `main` branch since this branch started, so this PR isn't valid before rebasing

File: src/lib/ug.ts (hunk 2 above)
Comment (me97esn) [reply src/lib/ug.ts:22]: So the first question you answered, `EqualityFilter:`, is this Typescript or Javascript? I thought it was a property `EqualityFilter`, not a type? Sorry for noob questions :)

Comment (exacs) [reply src/lib/ug.ts:22]: NO. This is not TypeScript. It is a combination of "parameter renaming" and "default value":

```
{
  EqualityFilter: filter = ""
}
```

means:

- that you should pass a property called `EqualityFilter` to the function
- that property internally is called `filter` in the function
- if no `EqualityFilter` parameter is passed, `""`  will be the default one

Comment (jhsware) [reply src/lib/ug.ts:22]: @exacs @me97esn It turns out that my original answer was correct. It is NOT parameter renaming, because this wasn't an object destructure operation. `EqualityFilter` is in fact an input parameter type. We got confused when reviewing because we only saw this diff snippet. The function looks like this:

`
async function ldapSearch({
  base = "OU=UG,DC=ug,DC=kth,DC=se",
  EqualityFilter: filter = "",
  attributes = [],
  scope = "sub",
  timeLimit = 10,
  ...extraOptions
}) {
  try {
    const options = {
      scope,
      timeLimit,
      filter,
      attributes,
      ...extraOptions,
    };

    const { searchEntries } = await ldapClient.search(base, options);
    return searchEntries;
  } catch (err) {
    err.message = "Error in LPDAP: " + err.message;
    throw err;
  }
}
`

Now it is more obvious, `filter`:
- is an optional parameter
- has the type `EqualityFilter`
- has the default value `filter = ""` if none was provided

**Comment by jhsware** [comment]
Force pushed due to rebase on main (rebase had zero diff to previous version)

File: src/lib/courseRoundsUtils.ts [hunk 6]
```diff
@@ -104,10 +105,11 @@ function shouldHaveAntagna(round) {
   const THREE_DAYS = 3 * 24 * 60 * 60 * 1000;
   const startDate = new Date(createStartDate(round));
 
-  return today() - startDate < THREE_DAYS;
+  // TS requires us to call .valueOf() on date obj
+  return today().valueOf() - startDate.valueOf() < THREE_DAYS;
```
Comment (jhsware) [inline src/lib/courseRoundsUtils.ts:109]: @me97esn this was the explanation. I had no choice...

File: src/lib/ug.ts (hunk 2 above)
Comment (jhsware) [reply src/lib/ug.ts]: Fixed in e622ecd

File: src/lib/utils.ts (hunk 1 above)
Comment (jhsware) [reply src/lib/utils.ts:52]: You always get inferred types and TS will find typos at compile time (we had such a bug in KTH Import Exams, I think it was a missing h in .length). But also this is a baseplate to start writing typed code moving forward. There is a commit added to this PR where types actually provide improved discovery of how to use the code.
